File: Parcial1/Melin.f90
PROGRAM MELIN
  IMPLICIT NONE

  ! ── Parámetros ────────────────────────
  INTEGER, PARAMETER :: N   = 10
  INTEGER, PARAMETER :: NPT = 600
  REAL(8), PARAMETER :: H    = 0.050D0
  REAL(8), PARAMETER :: ALFA = 0.2D0
  REAL(8), PARAMETER :: ANCHO = 1.0D0

  ! ── Variables ─────────────────────────
  REAL(8) :: Y(N), DYDT(N)
  REAL(8) :: T, DX
  REAL(8) :: t_inicio, t_fin, t_ejecucion
  INTEGER :: I

  DX = ANCHO / N
  T  = 0.0D0

  OPEN(UNIT=2, FILE='MELIN.dat', STATUS='REPLACE')

  ! ── Condición inicial ──────────────────
  DO I = 2, N
    Y(I) = 100.0D0
  END DO

  ! ── Condición de frontera en X=0 ───────
  Y(1) = 0.0D0

  ! ── Cronómetro inicia ──────────────────
  CALL CPU_TIME(t_inicio)

  CALL SOL(T, Y, DYDT, H, N, NPT, ALFA, DX)

  ! ── Cronómetro termina ─────────────────
  CALL CPU_TIME(t_fin)
  t_ejecucion = t_fin - t_inicio

  CLOSE(2)

  WRITE(*,'(A)')         '================================='
  WRITE(*,'(A,F10.6,A)') '  Tiempo de ejecucion: ', t_ejecucion, ' segundos'
  WRITE(*,'(A)')         '================================='

  STOP

CONTAINS

  ! =============================================
  SUBROUTINE DERIVS(X, Y, DYDT, ALFA, DX, N)
    IMPLICIT NONE
    INTEGER,  INTENT(IN)  :: N
    REAL(8),  INTENT(IN)  :: X, ALFA, DX
    REAL(8),  INTENT(IN)  :: Y(N)
    REAL(8),  INTENT(OUT) :: DYDT(N)

    REAL(8) :: ALFA2, DX2, FR_DER
    INTEGER :: I

    ALFA2   = ALFA * ALFA
    DX2     = DX * DX
    DYDT(1) = 0.0D0

    ! Puntos interiores
    DO I = 2, N-1
      DYDT(I) = ALFA2 * (Y(I+1) - 2.0D0*Y(I) + Y(I-1)) / DX2
    END DO

    ! Frontera aislada x=l  →  Neumann dT/dx=0
    FR_DER  = (4.0D0*Y(N) - Y(N-1)) / 3.0D0
    DYDT(N) = ALFA2 * (FR_DER - 2.0D0*Y(N) + Y(N-1)) / DX2

  END SUBROUTINE DERIVS

  ! =============================================
  SUBROUTINE RK4(Y, DYDT, N, X, H, YOUT, ALFA, DX)
    IMPLICIT NONE
    INTEGER, INTENT(IN)  :: N
    REAL(8), INTENT(IN)  :: Y(N), DYDT(N)
    REAL(8), INTENT(IN)  :: X, H, ALFA, DX
    REAL(8), INTENT(OUT) :: YOUT(N)

    REAL(8) :: YT(N), DYT(N), DYM(N)
    REAL(8) :: HH, H6, XH
    INTEGER :: I

    HH = H * 0.5D0
    H6 = H / 6.0D0
    XH = X + HH

    ! ── Paso 1 ────────────────────────────
    DO I = 1, N
      YT(I) = Y(I) + HH * DYDT(I)
    END DO
    CALL DERIVS(XH, YT, DYT, ALFA, DX, N)

    ! ── Paso 2 ────────────────────────────
    DO I = 1, N
      YT(I) = Y(I) + HH * DYT(I)
    END DO
    CALL DERIVS(XH, YT, DYM, ALFA, DX, N)

    ! ── Paso 3 ────────────────────────────
    DO I = 1, N
      YT(I)  = Y(I) + H * DYM(I)
      DYM(I) = DYT(I) + DYM(I)
    END DO
    CALL DERIVS(X+H, YT, DYT, ALFA, DX, N)

    ! ── Combinación final RK4 ─────────────
    DO I = 1, N
      YOUT(I) = Y(I) + H6 * (DYDT(I) + DYT(I) + 2.0D0*DYM(I))
    END DO

  END SUBROUTINE RK4

  ! =============================================
  SUBROUTINE SOL(T, Y, DYDT, H, N, NPT, ALFA, DX)
    IMPLICIT NONE
    INTEGER, INTENT(IN)    :: N, NPT
    REAL(8), INTENT(INOUT) :: T, Y(N), DYDT(N)
    REAL(8), INTENT(IN)    :: H, ALFA, DX

    REAL(8) :: YOUT(N), FR_DER
    INTEGER :: I, J

    ! ── Primer paso ───────────────────────
    FR_DER = (4.0D0*Y(N) - Y(N-1)) / 3.0D0
    WRITE(2,'(15F10.4)') T, (Y(I), I=1,N), FR_DER

    CALL DERIVS(T, Y, DYDT, ALFA, DX, N)
    CALL RK4(Y, DYDT, N, T, H, YOUT, ALFA, DX)

    FR_DER = (4.0D0*YOUT(N) - YOUT(N-1)) / 3.0D0
    WRITE(2,'(15F10.4)') T+H, (YOUT(I), I=1,N), FR_DER
    T = T + H

    ! ── Ciclo principal ───────────────────
    DO J = 1, NPT
      T = T + H
      DO I = 1, N
        Y(I) = YOUT(I)
      END DO
      CALL DERIVS(T, Y, DYDT, ALFA, DX, N)
      CALL RK4(Y, DYDT, N, T, H, YOUT, ALFA, DX)
      FR_DER = (4.0D0*YOUT(N) - YOUT(N-1)) / 3.0D0
      WRITE(2,'(15F10.4)') T, (YOUT(I), I=1,N), FR_DER
    END DO

  END SUBROUTINE SOL

END PROGRAM MELIN

File: Parcial1/Melin.cpp
// ============================================================
//  MELIN.cpp  —  Método de Líneas para la ecuación de difusión
//  Equivalente exacto del Melin.f90, con:
//    · Cronómetro de alta resolución (std::chrono)
//    · Salida MELIN_cpp.dat compatible con el notebook Python
// ============================================================
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <chrono>
#include <iomanip>

// ── Parámetros del problema ──────────────────────────────────
constexpr int    N     = 10;      // divisiones espaciales
constexpr int    NPT   = 600;     // pasos de tiempo a guardar
constexpr double H     = 0.050;   // paso de integración Δt
constexpr double ALFA  = 0.2;     // difusividad térmica α
constexpr double ANCHO = 1.0;     // longitud de la pared [m]

// ── Subroutine DERIVS ────────────────────────────────────────
void derivs(const std::vector<double>& Y,
            std::vector<double>&       DYDT,
            double alfa, double dx)
{
    const double a2  = alfa * alfa;
    const double dx2 = dx * dx;

    DYDT[0] = 0.0;

    for (int i = 1; i < N - 1; ++i)
        DYDT[i] = a2 * (Y[i+1] - 2.0*Y[i] + Y[i-1]) / dx2;

    double FR_DER = (4.0*Y[N-1] - Y[N-2]) / 3.0;
    DYDT[N-1] = a2 * (FR_DER - 2.0*Y[N-1] + Y[N-2]) / dx2;
}

// ── Subroutine RK4 ───────────────────────────────────────────
void rk4(const std::vector<double>& Y,
         const std::vector<double>& DYDT,
         double t, double h,
         std::vector<double>&       YOUT,
         double alfa, double dx)
{
    const double hh = h * 0.5;
    const double h6 = h / 6.0;

    std::vector<double> YT(N), DYT(N), DYM(N);

    for (int i = 0; i < N; ++i) YT[i] = Y[i] + hh * DYDT[i];
    derivs(YT, DYT, alfa, dx);

    for (int i = 0; i < N; ++i) YT[i] = Y[i] + hh * DYT[i];
    derivs(YT, DYM, alfa, dx);

    for (int i = 0; i < N; ++i) {
        YT[i]  = Y[i] + h * DYM[i];
        DYM[i] = DYT[i] + DYM[i];
    }
    derivs(YT, DYT, alfa, dx);

    for (int i = 0; i < N; ++i)
        YOUT[i] = Y[i] + h6 * (DYDT[i] + DYT[i] + 2.0*DYM[i]);
}

// ── Escribir fila en el .dat ─────────────────────────────────
void escribir_fila(std::ofstream& archivo,
                   double t,
                   const std::vector<double>& Y)
{
    double FR_DER = (4.0*Y[N-1] - Y[N-2]) / 3.0;
    archivo << std::fixed << std::setprecision(4)
            << std::setw(10) << t;
    for (int i = 0; i < N; ++i)
        archivo << std::setw(10) << Y[i];
    archivo << std::setw(10) << FR_DER << "\n";
}

// ── Main ─────────────────────────────────────────────────────
int main()
{
    const double DX = ANCHO / N;

    std::vector<double> Y(N), DYDT(N), YOUT(N);

    for (int i = 1; i < N; ++i) Y[i] = 100.0;
    Y[0] = 0.0;

    std::ofstream dat("MELIN_cpp.dat");
    if (!dat.is_open()) {
        std::cerr << "Error: no se pudo crear MELIN_cpp.dat\n";
        return 1;
    }

    double T = 0.0;

    auto t_inicio = std::chrono::high_resolution_clock::now();

    escribir_fila(dat, T, Y);

    derivs(Y, DYDT, ALFA, DX);
    rk4(Y, DYDT, T, H, YOUT, ALFA, DX);
    T += H;
    escribir_fila(dat, T, YOUT);

    for (int j = 1; j < NPT; ++j) {
        for (int i = 0; i < N; ++i) Y[i] = YOUT[i];
        T += H;
        derivs(Y, DYDT, ALFA, DX);
        rk4(Y, DYDT, T, H, YOUT, ALFA, DX);
        escribir_fila(dat, T, YOUT);
    }

    auto t_fin = std::chrono::high_resolution_clock::now();
    double t_ejecucion =
        std::chrono::duration<double>(t_fin - t_inicio).count();

    dat.close();

    std::cout << "\n=========================================\n";
    std::cout << "  MELIN C++ — Ecuación de difusión 1D\n";
    std::cout << "-----------------------------------------\n";
    std::cout << "  α      = " << ALFA  << "\n";
    std::cout << "  l      = " << ANCHO << " m\n";
    std::cout << "  N      = " << N     << "\n";
    std::cout << "  Δx     = " << DX    << " m\n";
    std::cout << "  h      = " << H     << " s\n";
    std::cout << "  Pasos  = " << NPT   << "\n";
    std::cout << "  t_fin  = " << T     << " s\n";
    std::cout << "-----------------------------------------\n";
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "  Tiempo de ejecucion: "
              << t_ejecucion << " segundos\n";
    std::cout << "  Archivo generado   : MELIN_cpp.dat\n";
    std::cout << "=========================================\n\n";

    return 0;
}